src/mono/wasm/runtime/binding_support.js

```diff
@@ -162,49 +160,174 @@ var BindingSupportLib = {
+				this.call_method (this.setup_js_cont, null, "mo", [ root.value, cont_obj ]);
+				
+				// register for GC of the Task after the JS side is done with the promise
+				this._js_owned_object_registry.register(result, gc_handle);
+
+				// register for instance reuse
+				this._js_owned_object_table.set(gc_handle, new WeakRef(result));
 			}
+
 			return result;
 		},
 
-		_js_owned_object_finalized: function (id) {
-			// The JS function associated with this ID has been collected by the JS GC.
-			// As such, it's not possible for this ID to be invoked by JS anymore, so
-			//  we can release the tracking weakref (it's null now, by definition),
-			//  and tell the C# side to stop holding a reference to the managed delegate.
-			this._js_owned_object_table.delete(id);
-			this.release_js_owned_object_by_handle(id);
+		_unbox_ref_type_root_as_object: function (root) {
+			this.bindings_lazy_init ();
+			if (root.value === 0)
+				return null;
+
+			// this could be JSObject proxy of a js native object
+			var js_handle = this._get_js_id (root.value);
+			if (js_handle > 0)
+				return this.mono_wasm_require_handle(js_handle);
+			// otherwise this is C# only object
+	
+			// get strong reference to Object
+			const gc_handle = this._alloc_gc_handle(root.value);
+
+			// see if we have js owned instance for this gc_handle already
+			var result = this._lookup_js_owned_object(gc_handle);
+
+			// If the JS object for this gc_handle was already collected (or was never created)
+			if (!result) {
+				result = {
+					// keep the gc_handle so that we could easily convert it back to original C# object for roundtrip
+					__js_owned_gc_handle__ : gc_handle
+				}
+	
+				// register for GC of the C# object after the JS side is done with the object
+				this._js_owned_object_registry.register(result, gc_handle);
+
+				// register for instance reuse
+				this._js_owned_object_table.set(gc_handle, new WeakRef(result));
+			}
+
+			return result;
+		},
+
+		_wrap_delegate_root_as_function: function (root) {
+			this.bindings_lazy_init ();
+			if (root.value === 0)
+				return null;
+
+			// get strong reference to the Delegate
+			const gc_handle = this._alloc_gc_handle(root.value);
+			return this._wrap_delegate_gc_handle_as_function(gc_handle);
+		},
+
+		_wrap_delegate_gc_handle_as_function: function (gc_handle) {
+			this.bindings_lazy_init ();
+
+			// see if we have js owned instance for this gc_handle already
+			var result = this._lookup_js_owned_object(gc_handle);
+
+			// If the function for this gc_handle was already collected (or was never created)
+			if (!result) {
+				// note that we do not implement function/delegate roundtrip
+				result = function() {
+					const delegateRoot = MONO.mono_wasm_new_root (BINDING.wasm_get_raw_obj(gc_handle, false));
+					try {
+						if (typeof result.__mono_delegate_invoke__ === "undefined")
+						result.__mono_delegate_invoke__ = BINDING.mono_wasm_get_delegate_invoke(delegateRoot.value);
```

Comment:
I don't think it needs to block merging of this PR, but I'd be much happier if we resolved this method before returning the delegate, so that the exception happens when wrapping the function instead of when it's called later. The point where it's being wrapped is more likely to have a useful stack trace so the developer can figure out what's going on.

Reply:
I thought we are doing it this way because we don't want to pay the price for delegates which are never called. Lazy.

Reply:
And fixed indentation.